mathjax3-ts/input/tex/color/ColorConfiguration.ts

```diff
@@ -0,0 +1,336 @@
+}
+
+/**
+ *  Get an RGB color
+ */
+function get_RGB(rgb: string) {
+    var rgbParts : string[] = rgb.replace(/^\s+/,'').replace(/\s+$/,'').split(/\s*,\s*/); 
+    var RGB = '#';
+
+    if (rgbParts.length !== 3)
+    {throw new TexError('ModelArg1', 'Color values for the %1 model require 3 numbers', 'RGB')}
+    
+    for (var i = 0; i < 3; i++) {
+        if (!rgbParts[i].match(/^\d+$/))
+        {throw new TexError('InvalidNumber', 'Invalid number')}
+        var n = parseInt(rgbParts[i]);
+        if (n > 255) {
+            throw new TexError('ModelArg2',
+                'Color values for the %1 model must be between %2 and %3',
+                'RGB', '0', '255');
+        }
+        var pN = n.toString(16); 
+        if (pN.length < 2) {pN = '0'+pN}
+        RGB += pN;
+    }
+    return RGB;
+}
+
+/**
+ *  Get a gray-scale value
+ */
+function get_gray(gray: string) {
+    if (!gray.match(/^\s*(\d+(\.\d*)?|\.\d+)\s*$/)) {
+        throw new TexError('InvalidDecimalNumber','Invalid decimal number')
+    }
+
+    var n : number = parseFloat(gray);
+    if (n < 0 || n > 1) {    
+        throw new TexError('ModelArg2',
+            'Color values for the %1 model must be between %2 and %3',
+            'gray', "0", "1");
+    }
+    var pn = Math.floor(n*255).toString(16); 
+    if (pn.length < 2) {
+        pn = '0' + pn;
+    }
+
+    return '#' + pn + pn + pn;
+}
+
+/**
+ *  Get a named value
+ */
+function get_named(name: string) : string {
+    if (COLORS.has(name)) {
+        return COLORS.get(name);
+    } else {
+        return name;
+    }
+}
+
+
+export let ColorMethods: Record<string, ParseMethod> = {};
+
+
+/**
+ * Override \color macro definition
+*/
+ColorMethods.Color = function(parser: TexParser, name: string) {
+    const model = parser.GetBrackets(name, "");
+    const color = getColor(model, parser.GetArgument(name));
+    
+    const math = parser.ParseArg(name);
+    
+    const node = parser.configuration.nodeFactory.create('node', 'mstyle', [math], {mathcolor: color});
+    parser.stack.env['color'] = color;
+    parser.Push(node);
+}
+
+ColorMethods.TextColor = function(parser: TexParser, name: string) {
+    const model = parser.GetBrackets(name, "");
+    const color = getColor(model, parser.GetArgument(name));
+    const old = parser.stack.env['color'];
+    parser.stack.env['color'] = color;
+    const math = parser.ParseArg(name);
+    if (old) {
+      parser.stack.env['color'];  // TODO: Volker, shouldn't we remove this?
+    } else {
+      delete parser.stack.env['color'];
+    }
+    const node = parser.configuration.nodeFactory.create('node', 'mstyle', [math], {mathcolor: color});
+    parser.Push(node);
+}
+
+/**
+ * Define the \definecolor macro
+*/
+ColorMethods.DefineColor =  function(parser: TexParser, name: string) {
+    var cname = parser.GetArgument(name),
+        model = parser.GetArgument(name),
+        def = parser.GetArgument(name);
+    
+    COLORS.set(cname, getColor(model, def));
```

Comment:
Effectively, we need a mechanism similar to the map handlers for parse definitions.
They allow to shadow definitions without actually overwriting them. Currently they are initialised creation `TeX` object creation, keeping the definitions local to each instance.
But if we want to have that possibility for custom structures to do something similiar, then we can either 
1. have an interface that these structures have to implement so they get initialised, or alternatively
2. each package configuration can bring it's own init method, that is called when a new `TeX` instance is created. 

The latter is probably easier and a bit more flexible. 
I'll have a think about it in the context of the `newcommand` and `ams` package (`DeclareMathOperator`) after PR #116 is merged.
We probably also need something similar for the `begingroup` package as well.

@OmarIthawi For now, just leave it like this. I am hoping to extend color to xcolor after you are done, which provides multiple named colour mappings. It might be easier to see if we can find a suitable abstraction then.

Reply:
@zorkow, I agree that 2 is probably the right choice, and that it can wait for now.  Just wanted it to be something that we keep in mind; as you point out, there are several other places that it may be important.